algorithms/src/main/kotlin/io/kinference/algorithms/gec/tokenizer/PreTrainedTokenizer.kt

```diff
@@ -0,0 +1,114 @@
+package io.kinference.algorithms.gec.tokenizer
+
+/**
+ * Base class for Tokenizer which using the vocabulary should implement following parameters
+ * @param vocabSize - length of Tokenizer vocabulary
+ * @param doLowerCase - boolean value for cased and uncased tokenizers
+ * @param unkToken - token which denotes unknown word
+ * @param sepToken - token which denotes end of sentence/end of segment
+ * @param padToken - token which denotes padding token for batch processing
+ * @param clsToken - token which denotes begin of sentence/begin of segment
+ * @param maskToken - token which denotes masking token for masked language modeling (not use in our approach)
+*/
+
+abstract class PreTrainedTokenizer(val doLowerCase: Boolean,
+                                   val unkToken: String,
+                                   val sepToken: String,
+                                   val padToken: String,
+                                   val clsToken: String,
+                                   val maskToken: String) {
+
+    abstract val vocabSize: Int
+
+    val unkId: Int by lazy { convertTokenToIdOnToken(unkToken) }
+    val sepId: Int by lazy { convertTokenToIdOnToken(sepToken) }
+    val padId: Int by lazy { convertTokenToIdOnToken(padToken) }
+    val clsId: Int by lazy { convertTokenToIdOnToken(clsToken) }
+    val maskId: Int by lazy { convertTokenToIdOnToken(maskToken) }
+
+
+    fun length(): Int {
+        return vocabSize
+    }
+
+    abstract fun tokenizeText(text: String): List<String>
```

Comment:
This function can probably be moved into the `Tokenizer` interface which will be used also by `BasicTokenizer`.